gms/gossiper.cc

```diff
@@ -842,6 +842,19 @@ future<> gossiper::failure_detector_loop() {
                     return g.failure_detector_loop_for_node(node, generation_number, live_endpoints_version);
                 });
             });
+            utils::chunked_vector<inet_address> nodes_down;
+            std::sort(nodes.begin(), nodes.end());
+            std::sort(_live_endpoints.begin(), _live_endpoints.end());
+            std::set_difference(nodes.begin(), nodes.end(), _live_endpoints.begin(), _live_endpoints.end(), std::back_inserter(nodes_down));
+            if (!nodes_down.empty()) {
+                logger.debug("failure_detector_loop: previous_live_nodes={}, current_live_nodes={}, nodes_down={}",
+                        nodes, _live_endpoints, nodes_down);
+                co_await seastar::async([this, &nodes_down] {
```

Comment:
Theoretically, _live_endpoints could change again while we're in this `co_await`, and we will miss it.

A more solid way would be to keep `nodes` always valid, pointing to the last some snapshot of _live_endpoints, and detect removal when you get the new snapshot. This way all changes must be detected.

Reply:
Yes, indeed _live_endpoints could change in the middle. I fixed this in v3.